### PR TITLE
Add PodDisruptionBudget when reconciling RunWorkloads

### PR DESCRIPTION
**Pull request by akrishna90**
## Is there a related GitHub Issue?

#1389

## What is this change about?

Adds PodDisruptionBudget

## Does this PR introduce a breaking change?

No

## Acceptance Steps

See issue #1389

## Tag your pair, your PM, and/or team

@julian-hj

### CODE REVIEW

**Review by julian-hj — APPROVED** [review]
🥇

File: statefulset-runner/controllers/pdb.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const PdbMinAvailableInstances = "50%"
+
+type Updater struct {
```
Comment (georgethebeatle) [inline statefulset-runner/controllers/pdb.go:19]: This struct fully qualified name name is `controllers.Updater` which is too broad and ambiguous. Let's rename it to something more specific like for example `controllers.PDBUpdater`?

File: statefulset-runner/controllers/runworkload_controller.go [hunk 2]
```diff
@@ -138,6 +141,12 @@ func (r *RunWorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	err = r.pdbUpdater.Update(ctx, updatedStatefulSet)
+	if err != nil {
+		r.Log.Error(err, "Error when creating or updating pod disruption budget")
+		return ctrl.Result{}, err
```
Comment (georgethebeatle) [inline statefulset-runner/controllers/runworkload_controller.go:147]: There is no test for this error case

File: statefulset-runner/controllers/pdb.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		return errors.Wrap(err, "pdb-updated-failed-to-set-owner-ref")
+	}
+
+	err := c.client.Create(ctx, pdb)
+
+	if k8serrors.IsAlreadyExists(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "failed to create pod distruption budget")
```
Comment (georgethebeatle) [inline statefulset-runner/controllers/pdb.go:65]: We should be consistent with other places in korifi where we wrap errors and use `fmt.Errorf("%w")`. We should not bring the unneeded dependency to `github.com/pkg/errors` from eirini.

File: statefulset-runner/controllers/pdb.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			},
+		},
+		Spec: policyv1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+			Selector:     statefulSet.Spec.Selector,
+		},
+	}
+
+	if err := controllerutil.SetOwnerReference(statefulSet, pdb, scheme.Scheme); err != nil {
+		return errors.Wrap(err, "pdb-updated-failed-to-set-owner-ref")
```
Comment (georgethebeatle) [inline statefulset-runner/controllers/pdb.go:56]: Use fmt.Errorf instead?

File: statefulset-runner/controllers/pdb.go (hunk 1 above)
Comment (akrishna90) [reply statefulset-runner/controllers/pdb.go:19]: We copied the implementation and related tests from from `eirini-controller`. Your suggestion make sense. I'll do the necessary changes.

File: statefulset-runner/controllers/runworkload_controller.go [hunk 5]
```diff
@@ -77,22 +77,25 @@ const (
...
 }
 
 func NewRunWorkloadReconciler(c client.Client, scheme *runtime.Scheme, log logr.Logger) *RunWorkloadReconciler {
 	return &RunWorkloadReconciler{
-		Client: c,
-		Scheme: scheme,
-		Log:    log,
+		Client:     c,
+		Scheme:     scheme,
+		pdbUpdater: NewUpdater(c),
```
[the remark this answers is not in the document]
Comment (akrishna90) [reply statefulset-runner/controllers/runworkload_controller.go:89]: Agreed. Done

File: statefulset-runner/controllers/runworkload_controller.go (hunk 2 above)
Comment (akrishna90) [reply statefulset-runner/controllers/runworkload_controller.go:147]: Added

File: statefulset-runner/controllers/runworkload_controller.go [hunk 6]
```diff
@@ -194,7 +203,7 @@ func truncateString(str string, num int) string {
 }
 
 func GetStatefulSetName(runWorkLoad korifiv1alpha1.RunWorkload) string {
-	nameSuffix, err := util.Hash(fmt.Sprintf("%s-%s", runWorkLoad.Spec.GUID, runWorkLoad.Spec.Version))
+	nameSuffix, err := hash(fmt.Sprintf("%s-%s", runWorkLoad.Spec.GUID, runWorkLoad.Spec.Version))
 	if err != nil {
 		panic(errors.Wrap(err, "failed to generate hash"))
```
[the remark this answers is not in the document]
Comment (akrishna90) [reply statefulset-runner/controllers/runworkload_controller.go:208]: done

File: statefulset-runner/controllers/pdb.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		return nil
+	}
+
+	return errors.Wrap(err, "failed to create pod distruption budget")
+}
+
+func (c *Updater) deletePDB(ctx context.Context, statefulSet *appsv1.StatefulSet) error {
+	err := c.client.DeleteAllOf(ctx, &policyv1.PodDisruptionBudget{}, client.InNamespace(statefulSet.Namespace), client.MatchingFields{"metadata.name": statefulSet.Name})
+
+	return errors.Wrap(err, "failed to delete pod distruption budget")
```
[the remark this answers is not in the document]
Comment (akrishna90) [reply statefulset-runner/controllers/pdb.go:71]: fixed

File: statefulset-runner/controllers/runworkload_controller.go [hunk 8]
```diff
@@ -463,3 +472,17 @@ func tcpSocketAction(runWorkload korifiv1alpha1.RunWorkload) *corev1.TCPSocketAc
...
 	return int32(millis / 1000) //nolint:gomnd
 }
+
+func hash(s string) (string, error) {
+	const MaxHashLength = 10
+
+	sha := sha256.New()
+
+	if _, err := sha.Write([]byte(s)); err != nil {
+		return "", errors.Wrap(err, "failed to calculate sha")
```
[the remark this answers is not in the document]
Comment (akrishna90) [reply statefulset-runner/controllers/runworkload_controller.go:482]: fixed

File: statefulset-runner/controllers/pdb.go (hunk 4 above)
Comment (akrishna90) [reply statefulset-runner/controllers/pdb.go:56]: fixed

File: statefulset-runner/controllers/pdb.go (hunk 3 above)
Comment (akrishna90) [reply statefulset-runner/controllers/pdb.go:65]: fixed

**Pull request merged**